Legend background and foreground colors can be modified

specmicp opened the pull request:
Resolves #2145 

- Change in visual properties of the legend do not require to get the handles anymore
- Background and foreground can be changed
- Defaults for these properties are taken from the settings
- The legend is also set to be draggable

specmicp commented:
It mostly works and is reasonable, except for the edge of the frame.

There are several options:

1. The edge color is the foreground color
2. No edge
3.The edge color can be freely modified by the user
4. inherit from the axes

I would tend to prefer 4, but it is not really set in the settings either. Maybe it should ?

The default is pretty good on black on white, but not that great for any other... It requires a bit of playing around

specmicp commented:
In this commit, I also set the legend to be draggable so it can be drag around so the user can do whatever he wants. That's quite useful (at least for me), but it might mess up with other redraw... @astrofrog  what do you think ?

astrofrog commented:
@specmicp - thanks for fixing this! Just to make sure I understand, what is the difference between options 1 and 4? The axes frame color is set to the foreground color anyway, right?

astrofrog commented:
The dragging is very nice by the way! I guess at the end of the day it depends if it causes all drawing-related tests to fail, but so far from playing around with it it works nicely. If it's a problem for redraws, you could allow location to be set to 'draggable' or something like that, so that it's opt-in?

astrofrog commented:
One small issue - I think maybe we should call it 'box' color not 'frame' color, otherwise it might be confusing:

![Screenshot from 2020-05-21 11-16-15](https://user-images.githubusercontent.com/314716/82549535-f9e03500-9b54-11ea-84aa-ff1ff99fb4f0.png)

astrofrog commented:
Thinking about the actual frame color a bit more, I think maybe having it be the same as the text (option 1) for now might make sense, and in future we could consider having separate global options for the text and frame colors for both axes and legend? (I'm not convinced this is needed)

However, it might be nice to have a toggle for turning the frame on/off on the legend?

astrofrog commented:
Sorry for all the comments, but I'm also wondering whether as we add more of these properties for the legend it might be worth having a dedicated State class for the legend, so that then options would be set with ``viewer.state.legend.frame_color`` instead of ``viewer.state.legend_frame_color``. It might be tidier to do this - what do you think? (could be done as a separate PR though!)

specmicp commented:
@ astrofrog 

> Just to make sure I understand, what is the difference between options 1 and 4? The axes frame color is set to the foreground color anyway, right?
> One small issue - I think maybe we should call it 'box' color not 'frame' color, otherwise it might be confusing:

frame is the property name for the box in the matplotlib API, that's why I went with that name, but I can see now how it can be confusing.
I changed it in the glue interface, not sure yet if I should change it in the code

I also added the option to draw or not the edge, and set up the edge color to the font color. We can play with it a bit before deciding. I agree that it does not make sense to allow full customization of the legend, because it's going to clutter the options for sure.

specmicp commented:
> The dragging is very nice by the way! I guess at the end of the day it depends if it causes all drawing-related tests to fail, but so far from playing around with it it works nicely. If it's a problem for redraws, you could allow location to be set to 'draggable' or something like that, so that it's opt-in?

I saw some strange things when resizing the plot, so I added the drag-behavior as an option. It's also clearer that it's a feature also !

specmicp commented:
> Sorry for all the comments, but I'm also wondering whether as we add more of these properties for the legend it might be worth having a dedicated State class for the legend, so that then options would be set with `viewer.state.legend.frame_color` instead of `viewer.state.legend_frame_color`. It might be tidier to do this - what do you think? (could be done as a separate PR though!)

I though of the same idea, so I think it's a good thing indeed.

astrofrog commented on glue/viewers/matplotlib/state.py:
```diff
@@ -34,13 +37,69 @@ def notify(self, *args, **kwargs):
+
+    visible = DeferredDrawCallbackProperty(False, docstring="Whether to show the legend")
+
+    loc_and_drag = DeferredDrawSelectionCallbackProperty(0, docstring="The location of the legend in the axis")
```
I think this would be clearer as just ``location`` as it was before - then the option can be ``draggable`` when it is draggable.

specmicp replied:
yes,

I changed back that name.
To have the argument to pass to matplotlib API, the property 'mpl_location' need to be used now

astrofrog commented on glue/viewers/matplotlib/state.py:
```diff
@@ -34,13 +37,69 @@ def notify(self, *args, **kwargs):
 VALID_WEIGHTS = ['light', 'normal', 'medium', 'semibold', 'bold', 'heavy', 'black']
 
-VALID_LOCATIONS = ['best',
+VALID_LOCATIONS = ['best (draggable)', 'best',
```
What about just ``'draggable'`` for the first option?

specmicp replied:
Much better indeed !

specmicp commented:
@astrofrog I change according to your request, and corrected some bugs. So I think this PR is ready to be merged for more testing.

I will do a new PR soon, with more tests and the documentation of the API, if no big problem is discovered soon.
